### PR TITLE
fix: federated auth with stream connectors

### PR DESCRIPTION
**Pull request by kian99**
This PR fixes an issue that prevented CLI users interacting with JAAS from running `juju debug-log`.

When running `juju debug-log` against a model that is hosted on a controller connected to JAAS, the operation hangs and nothing is returned. This is because JAAS proxies all client requests and did not implement the models `/log` endpoint. It also however has no way of authenticating the user.

Running `juju debug-log` does the following:
1. Connects to the Juju controller at `/model/<uuid>/api` creating an API client in the normal fashion, establishing a websocket connection and performing a login request.
2. It then goes on to establish a new separate connection to the `/model/<uuid>/log` endpoint, eventually establishing a websocket but instead of sending an RPC login request, it uses the apiClient's `ConnectStream` method where an HTTP basic auth header is sent that includes the user's password and includes the macaroons as cookies.
3. Proceeds to read from the websocket to stream logs.

With JAAS' introduction of login providers and particularly the session login provider, it was realised that there is no mechanism in place to send the session token used with OIDC login in the basic auth header used in the `ConnectStream` method. Because JAAS acts as a proxy layer for all controller and model level calls, JAAS has no way of authenticating requests to `/model/<uuid>/log`.

To fix this I have made the following changes in this PR.
1. ~~Add a `Token` method to the LoginProvider interface that returns a string. The string represents the token used for login.~~ After some discussion, this was changed to shift more logic into the login provider. Instead of returning a token that was then used as part of an HTTP header, the login provider interface now has an `AuthHeader` method that will return an HTTP header that can be used in places like the stream connector.
2. Added a `loginProvider` field to the `state` struct to keep the provider around, depending on how this is implemented, it's necessary to hold onto the provider because certain providers like the sessionTokenLoginProvider only obtain a login token after completing login.
3. I've replaced the logic in the `connectStream` method to use the login providers token rather than just the password from the state.

A benefit of this is that it further reduces the use of `state.password` ~~which could probably be removed with some further refactoring.~~ and now it completely removes the need for storing `tag`, `password`, `nonce`, and `macaroon` on the state struct since they are nicely contained within the login provider. 

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

- `make install`
- `juju bootstrap localhost test`
- `juju add-model test`
- `juju debug-log`

### CODE REVIEW

**Review by SimonRichardson — COMMENTED** [review]
@jameinel I'd like to see your thoughts around this one. We're doing a lot to auth to cater for JIMM, but we probably need to be 100% sure this is safe.

File: cmd/internal/loginprovider/tryinorderloginprovider.go
Comment (babakks) [inline cmd/internal/loginprovider/tryinorderloginprovider.go:34]: Maybe the same godoc here.

File: api/interface.go
Comment (alesstimec) [inline api/interface.go:192]: this could be in a separate `BasicAuthTokenProvider` interface and only `LoginProviders` where this method makes sense would implement it and in `api/state.go`, `LoginToken` method we can check if a `LoginProvider` implements the `BasicAuthTokenProvider` interface

Comment (alesstimec) [reply api/interface.go:192]: also naming suggestion: s/Token/AuthToken/

Comment (kian99) [reply api/interface.go:192]: All login providers should have this method, it's needed for a client to use any of the streaming endpoints. The way I refactored the code, the connection state doesn't need to hold onto the password anymore, it simply uses the login provider's token.
`userPassLoginProvider` - Token returns password.
`sessionTokenLoginProvider` - Token returns the session token.
`clientCredentialsLoginProvider` - Token returns client secret.
`tryInOrderLoginProvider` - Token returns the underlying provider's token.
The name I'm happy to change.

**Review by wallyworld — CHANGES_REQUESTED** [review]
There need to be unit tests added. eg the mock provider just returns "" to satisfy the new Token() interface. But there's no tests for tryinorder provider,  or any of the individual login providers, or the client itself.
The change seems simple enough - provide an abstraction to populate the auth header with a provider specific token, not just "password". But I'd like @tlm to look too.  We know we want to replace this infra with something OIDC compliant. Until we get there, are we ok with this change.

**Comment by kian99** [comment]
Thanks, I've added unit tests but I'm struggling to add a client test. 
I've got the workings of a test below in `api/apiclient_test.go`.
``` go
func (s *apiclientSuite) TestConnectControllerStreamAppliesLoginTokenHeader(c *gc.C) {
	catcher := api.UrlCatcher{}
	s.PatchValue(&api.WebsocketDial, catcher.RecordLocation)
	apiInfo := s.APIInfo(c)
	apiInfo.Tag = nil
	apiInfo.Password = ""
	apiInfo.SkipLogin = true
	sessionToken := "test-session-token"
	var output bytes.Buffer
	lp := api.NewSessionTokenLoginProvider(sessionToken, &output, nil)
	dialOpts := api.DialOpts{LoginProvider: lp}
	conn, err := api.Open(apiInfo, dialOpts)
	c.Assert(err, jc.ErrorIsNil)
	_, err = conn.ConnectControllerStream("/something", nil, nil)
	c.Assert(err, gc.ErrorMatches, "cannot use ConnectStream without logging in")
	expectedHeader := jujuhttp.BasicAuthHeader("", sessionToken)
	c.Assert(catcher.Headers().Get("Authorization"), gc.Equals, expectedHeader.Get("Authorization"))
}
```
This test requires that https://github.com/juju/juju/pull/17615 lands first, as it reverts moving the sessionToken login provider from the `cmd/internal`. 

The main issue though is that Juju doesn't implement any login method that accepts a session token so I can't use the session token login provider to actually login. And the result of that is `ConnectControllerStream` returns early with an error `cannot use ConnectStream without logging in`.
I can't patch the state's `loggedIn` field from a blackbox test. Which makes me think, there are already existing tests that verify the `connectStream` functionality. I'm purely trying to verify the header values on the request, that feels like it breaks the boundaries in what the surrounding tests are doing.

File: api/clientcredentialsloginprovider_test.go
Comment (tlm) [inline api/clientcredentialsloginprovider_test.go:80]: This defer should be straight after the error check.

File: api/clientcredentialsloginprovider_test.go
Comment (tlm) [inline api/clientcredentialsloginprovider_test.go:78]: c.Check please.

File: api/userpass_login_provider_test.go
Comment (tlm) [inline api/userpass_login_provider_test.go:21]: `Test` is not a valid test name. Please add a little more and even a comment above this func describing the behaviour we are trying to see exist.

File: cmd/internal/loginprovider/sessiontokenloginprovider.go
Comment (tlm) [inline cmd/internal/loginprovider/sessiontokenloginprovider.go:58]: Thanks for the comment but can we add some more depth to the comment explaining what this provider does when the func is called?

File: cmd/internal/loginprovider/tryinorderloginprovider.go
Comment (tlm) [inline cmd/internal/loginprovider/tryinorderloginprovider.go:34]: As above. Some more comments explaining the specifics of this implementation v satisfying an interface.

File: cmd/internal/loginprovider/tryinorderloginprovider.go
Comment (tlm) [reply cmd/internal/loginprovider/tryinorderloginprovider.go:35]: I am genuinely worried this is not sufficient for long running sessions or the possibility of re-use. In an OIDC world tokens are refreshed at a regular cadence. We aren't really handling that fact. We are also taking context away from the provider. Part of the call to the provider in the context that is has available is the time at which the information is requested.

If I was going to solve this I would probably store this as a closure that defer the call into the provider.

I am also noticing a theme that `Token` probably needs the opportunity to return an error to the call site. As two of the login providers have a complex state that needs to be setup and the call to `Token` can't be controlled. i.e what happens if this is called and the closure is nil? Does empty string mean that a token does or doesn't exist?

File: cmd/internal/loginprovider/tryinorderloginprovider_test.go
Comment (tlm) [inline cmd/internal/loginprovider/tryinorderloginprovider_test.go:34]: This test is asserting behaviour of a login provider that we haven't laid out in a contract yet. This feeds into the error comment above where we haven't really defined what the behaviours of these providers should be.

File: cmd/internal/loginprovider/tryinorderloginprovider_test.go
Comment (tlm) [inline cmd/internal/loginprovider/tryinorderloginprovider_test.go:36]: Please consider swapping most of these asserts over to Check's by where if there is a failure the test can keep running safely.

**Review by tlm — CHANGES_REQUESTED** [review]
Thanks for working on this. Have left an initial round of feedback. I am also looking into this flow a bit more tomorrow as a I think we may need some structural changes here.

The PR description mentions controller changes but I don't see them here?

File: cmd/internal/loginprovider/tryinorderloginprovider.go
Comment (kian99) [reply cmd/internal/loginprovider/tryinorderloginprovider.go:37]: I'm not sure I follow all your concerns here. Could you provide an example?

From I understand, you mean that the token returned could expire by the time the client uses it elsewhere? That is certainly possible if there was some client that performed an initial connection and then some hours later tried to make a call that required basic auth.

That to me speaks more to the problematic nature of the API and how one endpoint /api has a different authentication method versus /log. If they were the same there wouldn't be a need for `Token` as `Login` would always handle things properly.

As to your solution, I understand what you mean with the closure, the different providers can determine whether the token is still valid at the time the token is requested. The password for example would always be valid, the session token on the other hand should be able to be inspected to see if it is valid and if not restart the login flow.

But as it stands, I don't know that we this issue with long running sessions currently and we might never. Happy to refactor to something that caters to sessions if you think we should head in that direction.

File: cmd/internal/loginprovider/tryinorderloginprovider_test.go
Comment (kian99) [reply cmd/internal/loginprovider/tryinorderloginprovider_test.go:36]: I've changed the last one to a check. Not sure I agree the others should be checks. If there is a failure further up I don't see the value in continuing the test.

File: cmd/internal/loginprovider/tryinorderloginprovider_test.go
Comment (kian99) [reply cmd/internal/loginprovider/tryinorderloginprovider_test.go:34]: I hear you, how do we go about fixing that?

File: cmd/internal/loginprovider/tryinorderloginprovider.go
Comment (tlm) [reply cmd/internal/loginprovider/tryinorderloginprovider.go:38]: Yep, I understand that the likelihood of this being an issue given how short our login contexts are from the client. I still don't want to rely on that fact. Would rather get the design correct. If we can let the actual provider handle the call via closures it gives us full context inside the provider.

It also means the aggregate doesn't ever need to take on special case logic. Given my proposal above for introducing an error type that will be highly context aware.

File: cmd/internal/loginprovider/tryinorderloginprovider_test.go
Comment (tlm) [reply cmd/internal/loginprovider/tryinorderloginprovider_test.go:34]: Hopefully my change to the func signature propose above helps fix this.

**Review by tlm — COMMENTED** [review]
Thanks for the work on PR feedback. Please don't feel like what I am asking for is gospel. Happy to discuss.

File: api/interface.go
Comment (kian99) [reply api/interface.go]: Happy with this, I've added the extra return type and both suggested error consts. They are each only used in one place. I'll comment on them so we can discuss if they make sense.

File: cmd/internal/loginprovider/tryinorderloginprovider.go
Comment (kian99) [reply cmd/internal/loginprovider/tryinorderloginprovider.go]: Realised better what you meant after our discussion and added a closure around the underlying provider's token method. Let me know if this is sufficient.

File: api/apiclient.go
Comment (kian99) [inline api/apiclient.go:375]: I'm not a big fan of this logic but I believe it's needed because of the macaroon logic further below. The userpass login provider also handles macaroon login. During macaroon auth, the tag and password are both empty and I'm unsure what the server's behaviour is if the basic auth header is set and macaroons are included, there doesn't seem to be tests for the stream connector when using macaroons.

File: cmd/internal/loginprovider/tryinorderloginprovider.go
Comment (kian99) [inline cmd/internal/loginprovider/tryinorderloginprovider.go:32]: Here we set the default return result for the InOrderLoginProvider to the `ErrorLoginFirst` error. If the login is successful we replace this with a closure to the successful providers Token method.

File: api/sessiontokenloginprovider.go
Comment (kian99) [inline api/sessiontokenloginprovider.go:63]: @tlm Here is the other use of the new error types. I'm still basing the result off of the default type for a string but at least now the logic is encapsulated within the provider rather than leaking out into the rest of the code.
I could extend this to include a check for whether the existing token is still valid and return an error if it's not. Lmk what you think about that.

File: api/apiclient.go
Comment (wallyworld) [reply api/apiclient.go:375]: Will this work? Now that the underlying provider can return ErrorTokenNotAvailable, doesn't the error check need to account for that? And we'd avoid the `token != ""` check since we don't want to special case ""
But, ErrorTokenNotAvailable seems to imply a transient error whereas the logic here seems to want to make a decision based on a static decision - there is a token or there is not. We've introduced the error in this PR and nowhere is using it. So what's the plan for it?

File: api/clientcredentialsloginprovider.go
Comment (wallyworld) [inline api/clientcredentialsloginprovider.go:38]: Is clientSecret guaranteed to be not ""?
If not we'd need to return ErrorTokenNotAvailable

We have `NewClientCredentialsLoginProvider` above but there's no check to validate clientSecet (nor clientID).

File: api/clientcredentialsloginprovider.go
Comment (wallyworld) [inline api/clientcredentialsloginprovider.go:36]: returning

File: api/userpassloginprovider.go
Comment (wallyworld) [inline api/userpassloginprovider.go:63]: Also here - there's no check that password can't be set to "" and the contract of the Token() method would be violated

File: api/interface.go
Comment (wallyworld) [inline api/interface.go:197]: The description should include the errors returned under what circumstances, enclosed in `[ ]` for doc comment references.

And in so doing, we'd for example mention ErrTokenNtAvailable, but the code for some providers (see other comments) would then violate this contract.

**Review by wallyworld — COMMENTED** [review]
Some questions, plus exiting code to fix

`func (s *sessionTokenLoginProviderProviderSuite) Test(c *gc.C) {`

File: api/clientcredentialsloginprovider.go
Comment (kian99) [reply api/clientcredentialsloginprovider.go:38]: With the introduction of the error, the clientSecret could be "". There is not necessarily a problem with that, it's just based on whatever was provided. I don't think there is a meaningful check we could do in `NewClientCredentialsLoginProvider`, possibly we could prevent the user from providing an empty value if we wanted, I'm open to either way.

File: api/userpassloginprovider.go
Comment (kian99) [reply api/userpassloginprovider.go:63]: Right, with the change I mentioned above to avoid checking for `""` , I think it would be fine to allow empty strings here.

File: api/apiclient.go
Comment (kian99) [reply api/apiclient.go:375]: I'll make this change because I think it's the right thing regarding the contract that `LoginToken` is trying to form.

File: api/interface.go
Comment (wallyworld) [reply api/interface.go:197]: See other comment - the interface needs to define an expectation that is implemented consistently for every provider. This allows the caller / user of instances which satisfy the interface to know what they are getting and implement the logic accordingly. Otherwise we are breaking all sorts of abstraction / encapsulation concerns.

**Review by wallyworld — COMMENTED** [review]
Hopefully I've clarified my concerns.

File: api/clientcredentialsloginprovider.go
Comment (kian99) [reply api/clientcredentialsloginprovider.go:38]: To clarify my previous statement I've bolded the bits that I think are relevant,
> With the introduction of the error, the clientSecret **could be** "".

What I was trying to say there is that with the use of the error, the clientSecret can be an empty string if the provider deems that to be a valid value. Currently, the only time an error is returned is when a provider cannot provide a token until its `Login` method has been run - this applies to the sessionToken provider. The remaining providers are expected to return whatever string was used during creation of the provider, empty or not.

>If we want the semantics to be that we never return an empty value for the token

We no longer want to do that.

If you think we _should_ do that, that can be figured out. I'd also like to hear @tlm's thoughts on this.

File: api/apiclient.go
Comment (kian99) [inline api/apiclient.go:374]: Important bit here, we no longer optionally add the Basic Auth header.

File: api/apiclient.go
Comment (wallyworld) [inline api/apiclient.go:371]: The LoginToken method can return the `ErrorTokenNotAvailable` error. Is there a strategy for dealing with that by the caller given we just exit here on any error. What's the expected behaviour.

File: api/clientcredentialsloginprovider.go
Comment (wallyworld) [reply api/clientcredentialsloginprovider.go:38]: The Token() method should either always return a `ErrorTokenNotAvailable` error when the result string is "", or not. The API contract for an interface method must be implemented consistently across all implementations.

File: api/apiclient.go
Comment (kian99) [reply api/apiclient.go:364]: Currently the code that I've seen in Juju always creates a websocket connection to the model first before starting a stream to fetch logs. If you were to create an API connection and set `login: False` then start the stream you'd run into the `ErrorTokenNotAvailable` (which I've now replaced with `ErrLoginFirst`) so I think it's best left to the individual creating the connection to ensure the proper ordering is followed if they use the sessionTokenLoginProvider, I don't think we want to automatically perform a login request in order to obtain the session token.

File: api/userpassloginprovider.go
Comment (wallyworld) [inline api/userpassloginprovider.go:64]: At this point "tag" is no longer a Tag but a string. Maybe call it "tagStr" or even "userInfo" since really, that's what it is. I don't think "user" is suitable because the format would be "user-fred".

File: api/httpclient.go
Comment (wallyworld) [inline api/httpclient.go:96]: This whole block can be deleted.
Just pass `info.Tag` to `NewUserpassLoginProvider` directly.

File: api/interface.go
Comment (wallyworld) [inline api/interface.go:152]: Make the arg `tagStr` or something so it doesn't clash with `tag` below.

File: api/sessiontokenloginprovider_test.go
Comment (wallyworld) [inline api/sessiontokenloginprovider_test.go:108]: The original inline code cold have stayed. I liked it better.

**Review by wallyworld — APPROVED** [review]
Thanks very much for sticking with this. The new abstraction is much better.

File: api/userpassloginprovider.go
Comment (kian99) [reply api/userpassloginprovider.go]: I'll go with `userInfo`.

File: api/userpassloginprovider.go
Comment (kian99) [reply api/userpassloginprovider.go]: Changed.

File: api/httpclient.go
Comment (kian99) [reply api/httpclient.go]: Good catch

File: api/interface.go
Comment (kian99) [inline api/interface.go:153]: I'd like a second opinion on this. I added the extra check here `&&  tagStr != ""` because some tests were failing, e.g. `TestEndpointFiltering` where it seems to setup a fake controller that doesn't respond with a tag. So I'm not sure if this is the right solution.

Comment (wallyworld) [reply api/interface.go:153]: We should pass in a tag here. In hindsight after seeing this we should not have changed the p.tag to p.userInfo. If we need to ultimately pass around a tag, leave the tag on the provider and adjust that logic in there. Round tripping between a tag and string is not what we'd want to do. Then all this logic here can be reverted.

Comment (kian99) [reply api/interface.go]: Good point, I've done that and added some tests to verify things work if a nil Tag is passed in.

File: api/legacyloginprovider.go
Comment (kian99) [inline api/legacyloginprovider.go:1]: Before this lands I'd like to make sure folks are happy with the name change from `userpassloginprovider` to `legacyloginprovider` because I don't know if everyone would agree that this is "legacy" from Juju's perspective. I expect @SimonRichardson you might have thoughts on this.

**Review by tlm — APPROVED** [review]

**Pull request merged**
